### DISCUSSION TITLE
Terramate in CI with change detection - doable?

### DISCUSSION THREAD

**Discussion by macalac** [terramate-io/terramate#950] [Q&A] [answered]
Hey team, thank you very much for working on this project - a big fan of your work!

I wonder if you could help me with this issue. I am trying to setup [Multiple Terraform Plan reports](https://docs.gitlab.com/ee/user/infrastructure/iac/mr_integration.html#multiple-terraform-plan-reports) and pass state files between jobs in GitLab. I didn't manage to get this far as I got stuck on passing artefacts between the build and the deploy stages.

So far my CI appears as follows, I am uncertain about how to pass the path to plan files to the `artifacts` component. 
```
stage:
  - build
script:
  - terramate list --changed --tags prod --why
  - terramate run --changed --tags prod terraform plan -out "planfile"
artifacts:
####### PART I want to achieve
  paths:
    - stacks/prod/domain.com/file.plan
    - stacks/prod/domain.tech/file.plan
stage:
  - deploy

####### and here is the `apply` job, where it would even get more complicated,
####### How could I run this for each file?
script:
  - terramate run --changed --tags prod terraform apply <FILE_PLAN_PATH>
 ```

I dabbled with the use of eval but not sure where to go with it.
```
$ terramate run -- terramate experimental eval 'terramate.stack.path.absolute' | awk '{print $1"/file.plan"}'
/stacks/dev/dev.domain.com/file.plan
/stacks/prod/domain.com/file.plan
/stacks/prod/domain.tech/file.plan
```
I know I could probably work around it by generating separate CI parent-child pipelines per stack, but was hoping I can achieve the same with a single Terramate command.

Hope this question makes sense. Are there any examples you could point me to?

Thanks in advance.

**Comment by i4ki** [comment]
Hey @macalac 

Is there any problem with doing:
```
terramate run --changed --tags prod -- terraform apply file.plan
```
?

The `terramate run` chdir into the stack directory before executing the target command, then using relative path is okay.
Sorry if I misunderstand your question... if that's the case, can you elaborate?

**Comment by macalac** [comment]
Hey @i4ki and thanks for your time looking into this.

I didn't ask the right question I think because I didn't understand the problem :) Apologies.

I have a working pipeline and fixed my issue with passing artefacts by using globs
```
artifacts:
  paths:
    - **/file.plan
```

Now I am trying to save the `terraform show` results into a file. Instead of per stack, I'm getting them all appended in the same file in the root dir.
```
$ terramate run -- terraform plan -out file.plan
$ terramate run -- terraform show --json file.plan > file.plan.json
$ git status
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	stacks/prod/domain.com/file.plan
	stacks/prod/domain.tech/file.plan
	file.plan.json
```
Can I redirect output per stack, rather than in the root directory?

**Reply by switchdk** [reply, marked as answer]
Hi @macalac,
To answer your last question, whether it is possible to create an output per stack, you can do something like this:

```
terramate run -- terraform plan -out file.plan
terramate run -- bash -c "terraform show --json file.plan > file.plan.json"
```

I think the reason why it creates as single json plan from `terraform show` is because your shell is piping everything from your `terramate run` command into `file.plan.json`. 

I hope this gets you closer to your goal.

**Reply by macalac** [reply]
Nice one @switchdk , yeah that will make it definitely easier to work with, thank you.

I had a workaround in place but this will come in handy. Thanks!
